channels/sig_analog.c

```diff
@@ -1627,6 +1782,50 @@ void analog_handle_dtmf(struct analog_pvt *p, struct ast_channel *ast, enum anal
 		p->subs[idx].f.frametype = AST_FRAME_NULL;
 		p->subs[idx].f.subclass.integer = 0;
 		*dest = &p->subs[idx].f;
+	}  else if (p->callwaitdeluxepending) {
+		if (f->frametype == AST_FRAME_DTMF_END) {
+			unsigned int mssinceflash = ast_tvdiff_ms(ast_tvnow(), p->flashtime);
+			p->callwaitdeluxepending = 0;
+
+			/* This is the case where a user explicitly took action (made a decision)
+			 * for Call Waiting Deluxe.
+			 * Because we already handled the hook flash, if the user doesn't do
+			 * anything within a second, then we still need to eventually take
+			 * the default action (swap) for the call waiting.
+			 *
+			 * dahdi_write will also drop audio if callwaitdeluxepending is set HIGH,
+			 * and also check if flashtime hits 1000, in which case it will set the flag LOW and then take the
+			 * default action, e.g. analog_callwaiting_deluxe(p, 0);
+			 */
+
+			/* Slightly less than 1000, so there's no chance of a race condition
+			 * between do_monitor when it sees flashtime hitting 1000 and us. */
+			if (mssinceflash > 990) {
```

Comment:
Do we really want to be sensitive to timing?

Reply:
Yes, this is related to 1 second being the timeout specified for this feature.